File: go/cote/LTCD/두_수의_합.go
package LTCD

import (
	"fmt"
	"math/rand"
)

// const (
// 	TEST_CASE = 10
// 	N = 1000
// 	TARGET = 1.5 * N
// )

type 두_수의_합 struct {
	TEST_CASE int
	N int
	TARGET int
}

func (problem 두_수의_합) initialize(random *rand.Rand) []int {
	n := random.Intn(problem.N - 1) + 1
	ret := make([]int, n)

	set := make(map[int]bool, 0)

	for i :=0; i < n; i++ {
		
		for {
			c := random.Intn(problem.N)
			_, ok := set[c]
			if !ok {
				set[c] = true
				ret[i] = c
				break;
			}
		}
	}

	return ret 
}

func (problem 두_수의_합) solve(nums []int, target int) [][]int {
	num_to_index := make(map[int]int)
	ret := make([][]int, 0)

	for i, n := range nums {
		num_to_index[n] = i
	}

	for i, num := range nums {
		v, ok := num_to_index[target - num]
		if ok && i != v {
			ret = append(ret, []int{i, v})
		}
	}

	return ret
}

func (problem 두_수의_합) main() {
	random := rand.New(rand.NewSource(43))

	for it := 0; it < problem.TEST_CASE; it++ {
		nums := problem.initialize(random)
		target := random.Intn(problem.TARGET - 1) + 1
		fmt.Println(problem.solve(nums, target))
	}
}

File: python/cote/LTCD/두_수의_합.py
from typing import *
import random


TEST_CASES = 10
N = 1000
TARGET = 1.5 * N


def solve(nums: List[int], target: int) -> List[int]:
    num_to_index = {num: i for i, num in enumerate(nums)}

    for i, num in enumerate(nums):
        if target - num in num_to_index and i != num_to_index[target - num]:
            return [i, num_to_index[target - num]]

    return []


if __name__ == "__main__":

    random.seed(43)

    for _ in range(TEST_CASES):

        n = random.randrange(N - 1) + 1
        nums = random.sample(range(0, N - 1), n)
        target = random.randrange(1, TARGET - 1) + 1

        print(solve(nums, target))

    pass
